File: DotNet/Exercicios1/AutoResetEvt.cs
using System;
using System.Collections.Generic;

using System.Threading;
using Utils;

namespace Exercicios1 {
    public class AutoResetEvt {
        private Object monitor = new Object();
        private bool signaled; // event state
 
        private LinkedList<bool> waiters; // waiters list

        public AutoResetEvt(bool initialState) {
            signaled = initialState;
            waiters = new LinkedList<bool>();
        }

        public void Signal() {
            lock(monitor) {
                if (waiters.Count > 0) {
                    var wNode = waiters.First;
                   
                    waiters.RemoveFirst();
                    wNode.Value = true;  
                    Monitor.PulseAll(monitor);
                }
                else {
                    signaled = true;
                }
            }
        }

        public void PulseAll() {
            lock(monitor) {
                for (LinkedListNode<bool> it = waiters.First; it != null; it = it.Next)  
                    it.Value = true;
                waiters.Clear();
                Monitor.PulseAll(monitor);
            }
        }

        public bool  Wait(int timeout) { //  throws InterruptedException  
            lock(monitor) {
                // fast path

                // note that is signaled the list must be empty
                if (signaled) {
                    signaled = false;
                    return true;
                }
                if (timeout == 0) {
                    return false;
                }
                
                var wNode = waiters.AddLast(false);
                do {
                    try {
                        int refTime = Environment.TickCount;
                        Monitor.Wait(monitor, timeout);
                        if (wNode.Value == true) {
                            return true;
                        }
                        timeout = SynchUtils.RemainingTimeout(refTime, timeout);
                        if (timeout == 0) {
                            // abort wait
                            waiters.Remove(wNode);
                            return false;
                        }
                    }
                    catch (ThreadInterruptedException e) {
                        if (wNode.Value == true) {
                            // delay interruption and return success
                            Thread.CurrentThread.Interrupt();
                            return true;
                        }
                        // abort wait
                        waiters.Remove(wNode);
                        throw e;
                    }
                } while (true);
            }
        }
    }

}

File: Java/src/exercicios1/AutoResetEvt.java
package exercicios1;

import java.util.LinkedList;
import java.util.List;

import static utils.SynchUtils.INFINITE;
import static utils.SynchUtils.remainingTimeout;

/**
 * Created by jmartins on 01/10/2017.
 */
public class AutoResetEvt {
    private Object monitor = new Object();
    private boolean signaled; // event state

    // auxiliary class maintaining waiter state
    private static class Waiter {
        boolean done;
    }

    private List<Waiter> waiters; // waiters list

    public AutoResetEvt(boolean initialState) {
        signaled = initialState;
        waiters = new LinkedList<Waiter>();
    }

    public void signal() {
        synchronized(monitor) {
            if (waiters.size() > 0) {
                Waiter w = waiters.remove(0);
                w.done = true;
                monitor.notifyAll();
            }
            else {
                signaled= true;
            }
        }
    }

    public void pulseAll() {
        synchronized(monitor) {
            for(Waiter w : waiters)
                w.done = true;
            waiters.clear();
            monitor.notifyAll();
        }
    }

    public boolean await(long timeout) throws InterruptedException {
        synchronized(monitor) {
            // fast path
            // note that if it is signaled, the list must be empty
            if (signaled) {
                signaled = false;
                return true;
            }
            if (timeout == 0) {
                return false;
            }
            // prepare wait
            Waiter w = new Waiter();
            waiters.add(w);
            do {
                try {
                    long refTime = System.currentTimeMillis();
                    if (timeout == INFINITE)
                        monitor.wait();
                    else
                        monitor.wait(timeout);
                    if (w.done) {
                        return true;
                    }
                    timeout = remainingTimeout(refTime, timeout);
                    if (timeout == 0)  {
                        // abort wait
                        waiters.remove(w);
                        return false;
                    }

                }
                catch(InterruptedException e) {
                    if (w.done) {
                        // delay interruption and return success
                        Thread.currentThread().interrupt();
                        return true;
                    }
                    // abort wait
                    waiters.remove(w);
                    throw e;
                }
            } while (true);
        }
    }
}
